tagmedia

```diff
@@ -30,127 +48,481 @@ GetOptions(
+#
+sub usage
+{
+  (my $msg = <<"  = = = = = = = =") =~ s/^ {4}//mg;
+    Usage: tagmedia [OPTIONS] IMAGE
+    Add/remove tags to SUSE installation media.
+
+    IMAGE is an SUSE installation medium; either DVD image or disk image.
+
+    Options:
+
+      --show                    Show currently set tags (default if no option is given).
+      --digest DIGEST           Add digest DIGEST (md5, sha1, sha224, sha256, sha384, sha512).
+      --pad N                   Ignore N 2 kB blocks of padding (at image end).
+      --check                   Tell installer to run media check at startup.
+      --add-tag foo=bar         Add tag foo with value bar.
+      --remove-tag foo          Remove tag foo.
+      --clean                   Remove all tags.
+      --help                    Write this help text.
+
+  = = = = = = = =
+
+  print $msg;
+
+  exit shift;
+}
+
+
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# read_image_blob(image)
+#
+# Read first 36 kB.
+# Note: leaves $image_data ->{fh} open for calculate_digest() to continue reading.
+#
+# image: hash with image related data
+#
+sub read_image_blob
+{
+  my ($image) = @_;
+
+  my $blob_size = 0x9000;
+  $image->{blob_blocks} = $blob_size >> 9;		# in 0.5 kB units
+
+  die "$image->{name}: $!\n" unless open $image->{fh}, $image->{name};
+  die "$image->{name}: $!\n" unless sysread $image->{fh}, $image->{blob}, $blob_size;
+  die "$image->{name}: file too short\n" if $blob_size != length $image->{blob};
+}
+
+
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# get_image_type(image)
+#
+# Analyze image type and get some basic data.
+#
+# image: hash with image related data
+# image->{blob}: buffer containing a sufficiently large portion of the image (36 kB)
+#
+sub get_image_type
+{
+  my ($image) = @_;
+
+  # Get iso size from iso9660 header.
+  #
+  # Don't verify too much - most of the header might be gone. Checking for
+  # plausible size info should be enough.
+  #
+  my $iso_magic_ok = substr($image->{blob}, 0x8000, 8) eq "\x01CD001\x01\x00";
+  my $little = 4 * unpack("V", substr($image->{blob}, 0x8050, 4));
+  my $big = 4 * unpack("N", substr($image->{blob}, 0x8054, 4));
+  if(
+    $iso_magic_ok &&
+    $little &&
+    $little == $big &&
+    $little <= (-s $image->{name}) / 512
+  ) {
+    $image->{iso_blocks} = $little;
+  }
 
-if(defined($opt_pad)) {
-  if($opt_pad) {
-    unshift @opt_add_tag, "pad=$opt_pad";
-    if($pad_size >= $iso_size) {
-      die "$opt_pad: padding too big!\n";
+  # Scan mbr for last primary partition.
+  #
+  # Set $image->{part_start}, $image->{part_blocks} (in 0.5 kB units).
+  #
+  if(substr($image->{blob}, 0x1fe, 2) eq "\x55\xaa") {
+    for (my $idx = 0; $idx < 4; $idx++) {
+      my ($boot, $type, $start, $size) = unpack("Cx3Cx3VV", substr($image->{blob}, 0x1be + 0x10 * $idx, 0x10));
+      if(
+        $type &&
+        !($boot & 0x7f) &&
```

Comment:
can you document what do this check?

Reply:
in general here I think can help link to MBR description https://wiki.osdev.org/Partition_Table